File: packages/core-logic/src/compiler.ts
import { Rule, ProtectedRule } from './types.js';
import { createHash } from 'crypto';

/**
 * ADR 009: Compiled Rule Approach.
 * Rules are "compiled" into immutable runtime objects with upfront validation.
 */
export class RuleCompiler {
  static compile<T>(rule: Rule<T>): ProtectedRule<T> {
    // Upfront validation
    if (rule.priority < 0) {
      throw new Error(`Invalid priority for rule ${rule.id}`);
    }

    // Compute checksum for ADR 010
    const serialized = JSON.stringify({
      id: rule.id,
      priority: rule.priority,
      metadata: rule.metadata
    });
    const checksum = createHash('sha256').update(serialized).digest('hex');

    // Return immutable frozen object
    return Object.freeze({
      ...rule,
      checksum
    });
  }
}

File: packages/rust-fizzbuzz/src/compiler.rs
use crate::types::{Rule, ProtectedRule, RuleConfig};
use sha2::{Sha256, Digest};

pub struct RuleCompiler;

impl RuleCompiler {
    pub fn compile<T>(rule: Rule<T>) -> ProtectedRule<T> {
        let config = RuleConfig {
            id: rule.id.clone(),
            priority: rule.priority,
            metadata: rule.metadata.clone(),
        };
        
        let serialized = serde_json::to_string(&config).unwrap();
        let mut hasher = Sha256::new();
        hasher.update(serialized);
        let checksum = format!("{:x}", hasher.finalize());

        ProtectedRule {
            rule,
            checksum,
        }
    }
}
